addons/sale_coupon/models/sale_coupon_program.py

```diff
@@ -281,14 +281,13 @@ def _filter_programs_from_common_rules(self, order, next_order=False):
         """
         programs = self
         # Minimum requirement should not be checked if the coupon got generated by a promotion program (the requirement should have only be checked to generate the coupon)
+        # Product requirement should not be checked if the coupon got generated by a promotion program (the requirement should have only be checked to generate the coupon)
         if not next_order:
+            programs = programs and programs._filter_programs_on_products(order)
             programs = programs and programs._filter_on_mimimum_amount(order)
         programs = programs and programs._filter_on_validity_dates(order)
         programs = programs and programs._filter_unexpired_programs(order)
         programs = programs and programs._filter_programs_on_partners(order)
-        # Product requirement should not be checked if the coupon got generated by a promotion program (the requirement should have only be checked to generate the coupon)
-        if not next_order:
-            programs = programs and programs._filter_programs_on_products(order)
```

Comment:
If I understood correctly you change this because this specific client has a lot more programs per product than products per program, in which case this works fine, but I think the most general scenario (i.e. the most common case) is to have more products than programs, so this might actually deter performance in the general case